In [30]:
import pandas as pd

In [31]:
import numpy as np

удем осуществлять работу с непростым набором данных о состоянии здоровья лошадей, испытывающих кишечные колики. Цель – максимально корректно заполнить пропуски.

# Задание 
#### Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

In [79]:
columns = ['surgery?','Age', 'Hospital_Number', 'rectal_temperature', 'pulse', 'respiratory_rate', 'temperature_of_extremities', 'peripheral_pulse','mucous_membranes', 'capillary_refill_time', 'pain', 'peristalsis', 'abdominal_distension', 'nasogastric_tube', 'nasogastric_reflux', 'nasogastric_reflux_PH', 'rectal_examination-feces', 'abdomen', 'packed_cell_volume','total_protein', 'abdominocentesis_appearance', 'bdomcentesis_total_protein','outcome', 'urgical_lesion', '25', '26','27','cp_data' ]

In [80]:
horse_kolic = pd.read_csv('https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv', names=columns, delimiter=',', na_values=['?']) 

In [81]:


horse_kolic

surgery?  Age  Hospital_Number  rectal_temperature  pulse  \
0         2.0    1           530101                38.5   66.0   
1         1.0    1           534817                39.2   88.0   
2         2.0    1           530334                38.3   40.0   
3         1.0    9          5290409                39.1  164.0   
4         2.0    1           530255                37.3  104.0   
..        ...  ...              ...                 ...    ...   
295       1.0    1           533886                 NaN  120.0   
296       2.0    1           527702                37.2   72.0   
297       1.0    1           529386                37.5   72.0   
298       1.0    1           530612                36.5  100.0   
299       1.0    1           534618                37.2   40.0   

     respiratory_rate  temperature_of_extremities  peripheral_pulse  \
0                28.0                         3.0               3.0   
1                20.0                         NaN               NaN   
2                24.0                         1.0               1.0   
3                84.0                         4.0               1.0   
4                35.0                         NaN               NaN   
..                ...                         ...               ...   
295              70.0                         4.0               NaN   
296              24.0                         3.0               2.0   
297              30.0                         4.0               3.0   
298              24.0                         3.0               3.0   
299              20.0                         NaN               NaN   

     mucous_membranes  capillary_refill_time  ...  packed_cell_volume  \
0                 NaN                    2.0  ...                45.0   
1                 4.0                    1.0  ...                50.0   
2                 3.0                    1.0  ...                33.0   
3                 6.0                    2.0  ...                48.0   
4                 6.0                    2.0  ...                74.0   
..                ...                    ...  ...                 ...   
295               4.0                    2.0  ...                55.0   
296               4.0                    2.0  ...                44.0   
297               4.0                    1.0  ...                60.0   
298               3.0                    1.0  ...                50.0   
299               NaN                    NaN  ...                36.0   

     total_protein  abdominocentesis_appearance  bdomcentesis_total_protein  \
0              8.4                          NaN                         NaN   
1             85.0                          2.0                         2.0   
2              6.7                          NaN                         NaN   
3              7.2                          3.0                         5.3   
4              7.4                          NaN                         NaN   
..             ...                          ...                         ...   
295           65.0                          NaN                         NaN   
296            NaN                          3.0                         3.3   
297            6.8                          NaN                         NaN   
298            6.0                          3.0                         3.4   
299           62.0                          1.0                         1.0   

     outcome  urgical_lesion     25  26  27  cp_data  
0        2.0               2  11300   0   0        2  
1        3.0               2   2208   0   0        2  
2        1.0               2      0   0   0        1  
3        2.0               1   2208   0   0        1  
4        2.0               2   4300   0   0        2  
..       ...             ...    ...  ..  ..      ...  
295      3.0               2   3205   0   0        2  
296      3.0               1   2208   0   0        1  
297      2.0               1   3205   

In [82]:
horse_kolic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery?                     299 non-null    float64
 1   Age                          300 non-null    int64  
 2   Hospital_Number              300 non-null    int64  
 3   rectal_temperature           240 non-null    float64
 4   pulse                        276 non-null    float64
 5   respiratory_rate             242 non-null    float64
 6   temperature_of_extremities   244 non-null    float64
 7   peripheral_pulse             231 non-null    float64
 8   mucous_membranes             253 non-null    float64
 9   capillary_refill_time        268 non-null    float64
 10  pain                         245 non-null    float64
 11  peristalsis                  256 non-null    float64
 12  abdominal_distension         244 non-null    float64
 13  nasogastric_tube    

In [83]:
#Оставим только 8 столбцов для изучения


In [84]:
horse_kolic_study = horse_kolic [ ['surgery?', 'Age', 'rectal_temperature', 'pulse','respiratory_rate', 'temperature_of_extremities', 'pain', 'outcome'] ]

#### surgery?
1 = да, была операция
2 = лечили без операции

#### Age
1 = взрослая лошадь
2 = молодая лошадь (< 6 месяцев)

#### rectal_temperature
!!Повышенная температура может возникнуть из-за инфекции.
 температура может быть снижена, когда животное находится в позднем шоке
нормальная температура 37,8
 этот параметр обычно меняется по мере развития проблемы
например. может начаться нормально, затем стать повышенным из-за поражения, проходя обратно через нормальный диапазон, как лошадь впадает в шок
#### pulse           
является отражением состояния сердца: 30 - 40 нормально для взрослых.
редко бывает ниже нормы, хотя у спортивных лошадей
 может иметь частоту 20-25
 животные с болезненными поражениями или страдающие от циркуляторного шока могут иметь повышенную частоту сердечных сокращений     
#### respiratory_rate (частота дыхания)
Сомнительная для анализа из-за больших разбросов
нормальная частота составляет от 8 до 10 

#### temperature_of_extremities
1 = нормальное 
2 = теплое 
3 = прохладное 
4 = холодное 
холодные или прохладные конечности указывают на возможный шок
горячие конечности должны коррелировать с повышенной ректальной температурой. 

#### pain
боль – субъективная оценка уровня боли лошади 
 возможные значения:
1 = тревога, отсутствие боли 
2 = депрессия 
3 = прерывистая легкая боль 
4 = прерывистая сильная боль 
5 = постоянная сильная боль 
чем больнее, тем больше вероятность того, что потребуется хирургическое вмешательство.
предварительное лечение боли может в некоторой степени маскировать уровень боли 

#### outcome
1 = выжил 
2 = умер 
3 = подвергся эвтаназии

In [85]:
horse_kolic_study

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
3         1.0    9                39.1  164.0              84.0   
4         2.0    1                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0    1                 NaN  120.0              70.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature_of_extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           NaN   NaN      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         NaN   NaN      3.0  

[300 rows x 8 columns]

# Задание 2. Первичное изучение данных

Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

In [86]:
horse_kolic_study.describe()

surgery?         Age  rectal_temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

Из базовых статистик, рассчитаных выше по столбцу возраст видно, что максимальное значение = 9 , при этом, возможно лишь значение 1 или 2.
Рассмотрим данные строки с указанныим атрибутом возраста = 9: У большинства лошадей уже были операции, пульс выше среднего, что не является нормой для взрослых лошадей
Рассмотрим уникальные значения возраста. И увидим, что уникальных значений 2, это 1 и 9, значит была допущена ошибка человеческого фактора. Значение 9 = 2


In [88]:
#Рассмотрим уникальные значения возраста
horse_kolic_study['Age'].unique()

array([1, 9], dtype=int64)

In [97]:
#Выполним замену
horse_kolic_study.loc[(horse_kolic_study.Age ==9 ), 'Age' ] == 2

Series([], Name: Age, dtype: bool)

In [98]:
# теперь группы лошадей по возрасту распределены верно
horse_kolic_study['Age'].unique()

array([1, 2], dtype=int64)

In [99]:
#Найдем выбросы по значению пульса
Q1_pulse = horse_kolic_study['pulse'].quantile(0.25)
Q3_pulse = horse_kolic_study['pulse'].quantile(0.75)
IQR_H = Q3_pulse - Q1_pulse
IQR_H

40.0

In [100]:
lower_ = Q1_pulse - (1.5 * IQR_H)
higher_ = Q3_pulse + (1.5 * IQR_H)

In [103]:
#выведем уникальные значения по пульсу по найденным выбросам
horse_kolic_study[~horse_kolic_study['pulse'].between(lower_, higher_, inclusive=True)]['pulse'].unique()

array([164.,  nan, 150., 160., 184.])

симпатическая нервная система выделяет эпинефрин (адреналин), который пересиливает тонус блуждающего нерва и может быстро поднять частоту пульса лошади до 220-240 уд/мин.
Вследствие чего можно считать частоту пульса от 150 до 184 естественным выбросом

In [104]:
horse_kolic_study[~horse_kolic_study['pulse'].between(lower_, higher_, inclusive=True)].groupby('outcome').max()

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
outcome                                                               
1.0           2.0    2                39.2  150.0              72.0   
2.0           2.0    2                39.3  184.0              84.0   
3.0           2.0    1                39.5    NaN               NaN   

         temperature_of_extremities  pain  
outcome                                    
1.0                             3.0   5.0  
2.0                             4.0   5.0  
3.0                             3.0   3.0

Видим, что среди выживших лошадей есть лошади высоким пульсом 150, но они испитывали сильную постоянную боль(была операция)
Лошадь с максимальным пульсом 184 не выжила и так же испытывала сисльную постоянную боль(была операция).


In [57]:
#Посмотрим на выбросы по частоте дыхания
Q1_rate = horse_kolic_study['respiratory_rate'].quantile(0.25)
Q3_rate = horse_kolic_study['respiratory_rate'].quantile(0.75)
IQR_H = Q3_rate - Q1_rate
IQR_H

17.5

In [58]:
lower_rate = Q1_rate - (1.5 * IQR_H)
higher_rate = Q3_rate + (1.5 * IQR_H)

In [264]:
#Посмотрим на выбросы по частоте дыхания
horse_kolic_study[~horse_kolic_study['respiratory_rate'].between(lower_rate, higher_rate, inclusive=True)]['respiratory_rate'].min()

66.0

In [265]:
horse_kolic_study[~horse_kolic_study['respiratory_rate'].between(lower_rate, higher_rate, inclusive=True)]['respiratory_rate'].max()

96.0

Посмотрим на дата фрейм с данными по частете дыхания болше наименьшего значения выброса
Затем, посмотрим на средние значения для пульса и частоты дыхания в разрезе возраста и факта операции, для оценки выбросов.



In [411]:
horse_kolic_study[horse_kolic_study['respiratory_rate'] > 66].groupby(['Age','surgery?']).mean()

rectal_temperature       pulse  respiratory_rate  \
Age surgery?                                                     
1   1.0                38.416667   70.000000         82.000000   
    2.0                      NaN   56.000000         68.000000   
2   1.0                38.557143  139.714286         81.428571   
    2.0                39.000000  150.000000         72.000000   

              temperature_of_extremities      pain   outcome  
Age surgery?                                                  
1   1.0                         2.666667  2.333333  1.571429  
    2.0                         3.000000  3.000000  3.000000  
2   1.0                         2.000000  2.800000  1.428571  
    2.0                              NaN       NaN  1.000000

# Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [267]:
#Вспомним показатели основного дата фрейма

In [268]:
horse_kolic_study.describe()

surgery?         Age  rectal_temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.080000           38.167917   71.913043   
std      0.490305    0.271746            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

# rectal_temperature
Перейдем к заполнению пропусков
Рассмотрим rectal_temperature
Предположим, что температура зависит от temperature_of_extremities и pain

Попробуем выполнить группировку по переменной (скопируем для начала ДФ)

Мы не можем отрицать, что температура может изменяться в зависимости от температуры конечностей и состояния животного.
горячие конечности должны коррелировать с повышенной ректальной температурой а температура может быть снижена, когда животное находится в позднем шоке
Заполним пропуски для на основе вышесказанного:

In [413]:
horse_kolic_study_ = horse_kolic_study.copy()
horse_kolic_study_.groupby(['temperature_of_extremities', 'pulse']).median()

surgery?  Age  rectal_temperature  \
temperature_of_extremities pulse                                      
1.0                        36.0        1.5  1.0               37.20   
                           40.0        2.0  1.0               38.10   
                           42.0        2.0  1.0               38.25   
                           44.0        2.0  1.0               37.75   
                           46.0        2.0  1.0               38.60   
...                                    ...  ...                 ...   
4.0                        110.0       2.0  1.0               39.40   
                           112.0       1.0  1.0                 NaN   
                           120.0       1.0  1.0               38.85   
                           130.0       2.0  1.0               38.20   
                           164.0       1.0  2.0               39.10   

                                  respiratory_rate  pain  outcome  
temperature_of_extremities pulse                                   
1.0                        36.0               20.5   3.5      1.5  
                           40.0               18.0   2.0      1.0  
                           42.0               18.0   2.0      1.0  
                           44.0               20.0   3.0      1.0  
                           46.0                NaN   1.0      1.0  
...                                            ...   ...      ...  
4.0                        110.0              35.0   NaN      1.0  
                           112.0              13.0   2.0      1.0  
                           120.0              60.0   3.5      2.0  
                           130.0              16.0   2.0      3.0  
                           164.0              84.0   2.0      2.0  

[103 rows x 6 columns]

In [414]:
#выполним заполнение медианным значением
horse_kolic_study_['rectal_temperature'] = horse_kolic_study_['rectal_temperature'].fillna(horse_kolic_study_.groupby(['temperature_of_extremities', 'pulse'])['rectal_temperature'].transform('median'))

In [310]:
horse_kolic_study['rectal_temperature'].describe()

count    240.000000
mean      38.167917
std        0.732289
min       35.400000
25%       37.800000
50%       38.200000
75%       38.500000
max       40.800000
Name: rectal_temperature, dtype: float64

In [311]:
horse_kolic_study_['rectal_temperature'].describe()

count    267.000000
mean      38.192884
std        0.721388
min       35.400000
25%       37.800000
50%       38.200000
75%       38.600000
max       40.800000
Name: rectal_temperature, dtype: float64

# pulse
Знамем, что животные с болезненными поражениями или страдающие от циркуляторного шока могут иметь повышенную частоту сердечных сокращений, а значит может зависить от уровня боли
холодные или прохладные конечности указывают на возможный шок
На основе вышесказанного заполним пропуски по пульсу

In [313]:
#Заполним пропуски медианным значением
horse_kolic_study_['pulse'] = horse_kolic_study_['pulse'].fillna(horse_kolic_study_.groupby(['temperature_of_extremities', 'pain'])['pulse'].transform('median'))

In [314]:
horse_kolic_study_['pulse'].describe()

count    291.000000
mean      72.288660
std       28.161343
min       30.000000
25%       48.000000
50%       64.000000
75%       90.000000
max      184.000000
Name: pulse, dtype: float64

In [315]:
horse_kolic_study['pulse'].describe()

count    276.000000
mean      71.913043
std       28.630557
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: pulse, dtype: float64

### respiratory_rate
несмотря на большой разброс частоты дыхания, можно предположить, что частота дыхания зависит от пульса и состояния лошади
Рассчитаем среднее значение частоты дыхания в зависимости от этих показателей


In [415]:
horse_kolic_study_.groupby(['pain', 'pulse'])['respiratory_rate'].median()

pain  pulse
1.0   40.0     16.0
      42.0     14.0
      44.0     40.0
      45.0     16.0
      46.0      NaN
               ... 
5.0   100.0    30.0
      112.0    24.0
      120.0    28.0
      136.0    48.0
      150.0    50.0
Name: respiratory_rate, Length: 111, dtype: float64

In [416]:
horse_kolic_study_['respiratory_rate'] = horse_kolic_study_['respiratory_rate'].fillna(horse_kolic_study_.groupby(['pain', 'pulse'])['respiratory_rate'].transform('median'))

In [318]:
horse_kolic_study_['respiratory_rate'].describe()

count    279.000000
mean      30.430108
std       16.872393
min        8.000000
25%       20.000000
50%       27.000000
75%       36.000000
max       96.000000
Name: respiratory_rate, dtype: float64

In [319]:
horse_kolic_study['respiratory_rate'].describe()

count    242.000000
mean      30.417355
std       17.642231
min        8.000000
25%       18.500000
50%       24.500000
75%       36.000000
max       96.000000
Name: respiratory_rate, dtype: float64

## pain
pain чем больнее, тем больше вероятность того, что потребуется хирургическое вмешательство
rectal_temperature - парметр меняется по мере развития проблемы
Повышенная температура может возникнуть из-за инфекции.
температура может быть снижена, когда животное находится в позднем шоке

сформировав свод по параметрам факта операции и ректанольной температуры видим, что для непрооперированных лошадей уровень боли  повышен с низкой ректальной температурой, так как в состоянии шока она может снижаться.
У прооперированных лошадей уровень боли снижен с более высокой ректальной температрой.




In [321]:
horse_kolic_study_.groupby(['surgery?', 'rectal_temperature'])['pain'].median()

surgery?  rectal_temperature
1.0       35.4                  4.0
          36.4                  4.0
          36.5                  4.0
          36.8                  NaN
          37.0                  4.0
                               ... 
2.0       39.4                  NaN
          39.5                  3.0
          39.6                  2.0
          40.0                  2.0
          40.3                  2.0
Name: pain, Length: 66, dtype: float64

In [322]:
horse_kolic_study_['pain'] = horse_kolic_study_['pain'].fillna(horse_kolic_study_.groupby(['surgery?', 'rectal_temperature'])['pain'].transform('median'))

In [323]:
horse_kolic_study_['pain'].describe()

count    285.000000
mean       2.963158
std        1.262770
min        1.000000
25%        2.000000
50%        3.000000
75%        4.000000
max        5.000000
Name: pain, dtype: float64

In [324]:
horse_kolic_study['pain'].describe()

count    245.00000
mean       2.95102
std        1.30794
min        1.00000
25%        2.00000
50%        3.00000
75%        4.00000
max        5.00000
Name: pain, dtype: float64

# temperature_of_extremities
1 = нормальное
2 = теплое
3 = прохладное
4 = холодное
холодные или прохладные конечности указывают на возможный шок, зависит от
горячие конечности должны коррелировать с повышенной ректальной температурой.


In [327]:
horse_kolic_study_.groupby(['pulse', 'rectal_temperature'])['temperature_of_extremities'].median()

pulse  rectal_temperature
30.0   37.8                  NaN
       38.5                  NaN
36.0   37.2                  1.0
       38.0                  3.0
38.0   37.6                  3.0
                            ... 
150.0  38.8                  1.0
       39.0                  NaN
160.0  38.6                  3.0
164.0  39.1                  4.0
184.0  38.8                  1.0
Name: temperature_of_extremities, Length: 189, dtype: float64

In [328]:
horse_kolic_study_['temperature_of_extremities'] = horse_kolic_study_['temperature_of_extremities'].fillna(horse_kolic_study_.groupby(['rectal_temperature'])['temperature_of_extremities'].transform('median'))

In [329]:
horse_kolic_study['temperature_of_extremities'].describe()

count    244.000000
mean       2.348361
std        1.045054
min        1.000000
25%        1.000000
50%        3.000000
75%        3.000000
max        4.000000
Name: temperature_of_extremities, dtype: float64

In [330]:
horse_kolic_study_['temperature_of_extremities'].describe()

count    284.000000
mean       2.339789
std        1.023505
min        1.000000
25%        1.000000
50%        3.000000
75%        3.000000
max        4.000000
Name: temperature_of_extremities, dtype: float64

In [332]:
#Посмотрим на пропуск по surgery?
horse_kolic_study_[horse_kolic_study_['surgery?'].isnull()]

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
132       NaN    1                38.0   48.0              20.0   

     temperature_of_extremities  pain  outcome  
132                         3.0   4.0      NaN

In [418]:
#видим, что все показатели в номре, кроме уровня боли. предположим, что это молодая лошадь с прерывистой сильной болью
horse_kolic_study_['surgery?'] = horse_kolic_study_['surgery?'].fillna(2)

In [419]:
#Проверим наличие пропусков
horse_kolic_study_['surgery?'].isna().sum()

0

Заполнили все возможные зависимые показатели, для остальные нельзя просто так удалить
для заполнения пропусков заполним данными медианного значения по возрасту и была или нет операция
для значения pain


In [420]:
horse_kolic_study_['pain'] = horse_kolic_study_['pain'].fillna(horse_kolic_study_.groupby(['surgery?', 'Age' ])['pain'].transform('median'))

In [421]:
#заполним пропуски по пульсу в соответствии с уровнем боли
horse_kolic_study_['pulse'] = horse_kolic_study_['pulse'].fillna(horse_kolic_study_.groupby(['pain'])['pulse'].transform('median'))

In [422]:
horse_kolic_study_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    300 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal_temperature          267 non-null    float64
 3   pulse                       300 non-null    float64
 4   respiratory_rate            270 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        300 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [423]:
#заполним пропуски по частоте дыхания в соответствии с уровнем боли и пульсу
horse_kolic_study_['respiratory_rate'] = horse_kolic_study_['respiratory_rate'].fillna(horse_kolic_study_.groupby(['pain', 'pulse'])['respiratory_rate'].transform('mean'))

In [424]:
horse_kolic_study_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    300 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal_temperature          267 non-null    float64
 3   pulse                       300 non-null    float64
 4   respiratory_rate            293 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        300 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [425]:
#выполним заполнение медианным значением пропуски по ректальной температуре в зависимости от боли и пульса
horse_kolic_study_['rectal_temperature'] = horse_kolic_study_['rectal_temperature'].fillna(horse_kolic_study_.groupby(['pain','pulse'])['rectal_temperature'].transform('median'))

In [426]:
horse_kolic_study_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    300 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal_temperature          294 non-null    float64
 3   pulse                       300 non-null    float64
 4   respiratory_rate            293 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        300 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [427]:
#выполним заполнение медианным значением'temperature_of_extremities' в зависимости от пульса и 'rectal_temperature'
horse_kolic_study_['temperature_of_extremities'] = horse_kolic_study_['temperature_of_extremities'].fillna(horse_kolic_study_.groupby(['pulse', 'rectal_temperature'])['temperature_of_extremities'].transform('median'))

In [428]:
horse_kolic_study_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    300 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal_temperature          294 non-null    float64
 3   pulse                       300 non-null    float64
 4   respiratory_rate            293 non-null    float64
 5   temperature_of_extremities  269 non-null    float64
 6   pain                        300 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


Больше не можем заполнять в соответствии с зависимости от изменения ближайших показателей.
Предположим, что для исхода была предрасположенность по показателям rectal_temperature, respiratory_rate, temperature_of_extremities

In [429]:
# Выполним заполнение пропусков в соответствии с исходом
horse_kolic_study_['rectal_temperature'] = horse_kolic_study_['rectal_temperature'].fillna(horse_kolic_study_.groupby('outcome')['rectal_temperature'].transform('median'))


In [430]:
horse_kolic_study_['respiratory_rate'] = horse_kolic_study_['respiratory_rate'].fillna(horse_kolic_study_.groupby('outcome')['respiratory_rate'].transform('median'))


In [431]:
horse_kolic_study_['temperature_of_extremities'] = horse_kolic_study_['temperature_of_extremities'].fillna(horse_kolic_study_.groupby('outcome')['temperature_of_extremities'].transform('median'))

In [432]:
horse_kolic_study_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    300 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal_temperature          300 non-null    float64
 3   pulse                       300 non-null    float64
 4   respiratory_rate            300 non-null    float64
 5   temperature_of_extremities  300 non-null    float64
 6   pain                        300 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [433]:
horse_kolic_study_[horse_kolic_study_['outcome'].isnull()]

surgery?  Age  rectal_temperature  pulse  respiratory_rate  \
132       2.0    1                38.0   48.0              20.0   

     temperature_of_extremities  pain  outcome  
132                         3.0   4.0      NaN

In [436]:
#Предположим, что исход зависит от пульса и ректальной температуры 
horse_kolic_study_['outcome'] = horse_kolic_study_['outcome'].fillna(horse_kolic_study_.groupby(['rectal_temperature', 'pulse'])['outcome'].transform('mean'))

In [437]:
horse_kolic_study_.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    300 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal_temperature          300 non-null    float64
 3   pulse                       300 non-null    float64
 4   respiratory_rate            300 non-null    float64
 5   temperature_of_extremities  300 non-null    float64
 6   pain                        300 non-null    float64
 7   outcome                     300 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [438]:
horse_kolic_study_.isna().sum()

surgery?                      0
Age                           0
rectal_temperature            0
pulse                         0
respiratory_rate              0
temperature_of_extremities    0
pain                          0
outcome                       0
dtype: int64